woocommerce/woocommerce
# Add order attribution feature

### Submission Review Guidelines:


### Changes proposed in this Pull Request:
**_[♻️  Updated 24 Nov 2023]_**

This pull request adds the [WooCommerce Order Source Attribution](https://github.com/woocommerce/woocommerce-order-source-attribution) feature – now simply "Order Attribution" –  to WooCommerce core. This feature allows store owners attribute orders to a specific traffic source.

Prototype PT: pe2C5g-ur-p2
Design i1.5: pe2C5g-HR-p2

### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Install this PR branch
    1. Check out this branch
    2. Run `pnpm install` to install all of the dependencies.
    3. Build and install the file on a WordPress site:
        - Navigate to `plugins/woocommerce/` and run `pnpm run build:zip` to build the plugin zip file, then install it on a WordPress site.
        OR 
        - Navigate to `plugins/woocommerce/` and run `pnpm run build` to build the plugin, then symlink it to a WordPress site.
5. Activate the plugin. 
6. Substitute the checkout block for the old shortcode checkout `[woocommerce_checkout]`
7. Navigate to WooCommerce > Settings > Advanced > Features and ensure the "Order Source Attribution" feature is available, and enabled by default.
8. The Orders page will have a new column labeled "Origin" that will display the source of the order.
9. The Order Edit page will have 2 new meta boxes: Order information and Customer history. These boxes will have more details about the origin of the order, and the complete history for the customer of the order.
10. Place a new order via the shop with the admin user, and confirm that the values in the Orders table and Order Edit page are populated (Probably with the source `Direct`)
11. Open an incognito tab and navigate to twitter.com. 
12. Using the following bookmarklet/JS snippet in the dev console, create an auto-clicking link to your test store:
    <details>
    <summary>JavaScript Code</summary>

    ```js

    javascript: (function() {
        var url = prompt("Enter the URL:", "[WC Shop URL]");
        if (url) {
            var button = document.createElement("a");
            button.href = url;
            button.innerText = "Go to URL";
            button.style.position = "absolute";
            button.style.top = "0";
            button.style.left = "0";
            button.style.padding = "10px";
            button.style.backgroundColor = "#3498db";
            button.style.color = "#fff";
            button.style.textDecoration = "none";
            button.style.zIndex = "9999";
            document.body.appendChild(button);
            button.click();
        }
    })();
    ```
    </details> 

8. Complete a purchase as an anonymous customer.
9. Confirm that the Origin value is "Referral: Twitter.com" on the Orders table and Order edit page.

**Deep testing variations**
1. Test with and without HPOS to ensure the data is stored and presented correctly.
2. Test with both the Block Checkout and the Classic Checkout (not Shortcode).
3. Test with UTM, referrals from Google and other sites, direct type-ins.
4. Test API and wp-admin order creation continues to work.
5. Test with the order attribution feature disabled.


### Checkout Block Support

Support for th WooCommerce checkout block was added in the sub-PR #41592 

### Screenshots
<img width="813" alt="Settings page showing new Order Attribution feature" src="https://github.com/woocommerce/woocommerce/assets/228780/80da5f97-1e15-4c8f-ba8c-301c015ef13d">

<img width="2234" alt="Orders page showing Origin column" src="https://github.com/woocommerce/woocommerce/assets/871924/286732ec-74a4-4690-a307-bd3c79e4e88d">
<img width="294" alt="Order Edit page showing new metaboxes" src="https://github.com/woocommerce/woocommerce/assets/228780/9a461d83-a530-45b0-9c0d-34da0be2b2a1">



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
